python/mxnet/test_utils.py

```diff
@@ -79,10 +79,117 @@ def random_sample(population, k):
     return population_copy[0:k]
 
 
-def rand_sparse_ndarray(shape, stype, density=None):
-    """Generate a random sparse ndarray. Returns the ndarray, value(np) and indices(np) """
+def _get_uniform_dataset_csr(num_rows, num_cols, density=0.1):
+    """Returns CSRNDArray with uniform distribution
+    """
+    if num_rows <= 0 or num_cols <= 0:
+        raise ValueError("num_rows or num_cols should be greater than 0")
+
+    if density < 0 or density > 1:
+        raise ValueError("density has to be between 0 and 1")
+
+    return mx.nd.array(sp.rand(num_rows, num_cols, density).toarray())._to_csr()
+
+
+def _get_powerlaw_dataset_csr(num_rows, num_cols, density=0.1):
+    """Returns CSRNDArray with powerlaw distribution
+    with exponentially increasing number of non zeros in each row.
+    Not supported for cases where totalnnz < 2*num_rows. This is because
+    the algorithm first tries to ensure that there are rows with no zeros by
+    putting non zeros at beginning of each row.
+    """
+
+    def validate_inputs(totalnnz, num_rows, num_cols):
+        """Validate inputs"""
+        if num_rows <= 0 or num_cols <= 0:
+            raise ValueError("num_rows or num_cols should be greater than 0")
+
+        if density < 0 or density > 1:
+            raise ValueError("density has to be between 0 and 1")
+
+        if totalnnz < 2 * num_rows:
+            raise ValueError("not supported for this density: %s"
+                             " for this shape (%s,%s)"
+                             " Please keep :"
+                             " num_rows * num_cols * density >= 2 * num_rows"
+                             % (density, num_rows, num_cols))
+
+    totalnnz = int(num_rows * num_cols * density)
+    validate_inputs(totalnnz, num_rows, num_cols)
+
+    unusednnz = totalnnz
+    output_arr = np.zeros((num_rows, num_cols))
+    # Start with ones on each row so that no row is empty
+    for row in range(num_rows):
+        output_arr[row][0] = rnd.uniform(0.001, 1)
+        unusednnz = unusednnz - 1
+        if unusednnz <= 0:
+            return mx.nd.array(output_arr)._to_csr()
+
+    # Populate rest of matrix with 2^i items in ith row.
+    # if we have used all total nnz return the sparse matrix
+    # else if we reached max column size then fill up full columns unit we use all nnz
+    col_max = 2
+    for row in range(num_rows):
+        col_limit = min(num_cols, col_max)
+        # In case col_limit reached assign same value to all elements, which is much faster
+        if col_limit == num_cols and unusednnz > col_limit:
+            output_arr[row] = rnd.uniform(0.001, 1)
+            unusednnz = unusednnz - col_limit + 1
+            if unusednnz <= 0:
+                return mx.nd.array(output_arr)._to_csr()
+            else:
+                continue
+        for col_index in range(1, col_limit):
+            output_arr[row][col_index] = rnd.uniform(0.001, 1)
```

Comment:
It will reduce from  52 seconds to around 30 seconds for 128 * 16M with density 0.65. For smaller densities < 0.1 the nnzs are lesser and the gap in performance improvement will also be lesser.